File: datastructures/bst/binary_search_tree.py
class Node:
    def __init__(self, key, val, left=None, right=None, n=0):
        self.key = key
        self.val = val
        self.left = left
        self.right = right
        self.n = n


class BinarySearchTree:
    def __init__(self):
        self.root = None

    def size(self, node):
        if not node:
            return 0
        return node.n

    def put(self, key, val):
        self.root = self._put(self.root, key, val)

    def _put(self, node, key, val):
        if not node: return Node(key, val, n=1)

        if node.key == key:
            node.val = val
        elif key < node.key:
            node.left = self._put(node.left, key, val)
        else:
            node.right = self._put(node.right, key, val)

        node.N = self.size(node.left) + self.size(node.right) + 1
        return node

    def get(self, key):
        return self._get(self.root, key)

    def _get(self, node, key):
        if not node: return -1
        if node.key == key:
            return node.val
        elif key < node.key:
            return self._get(node.left, key)
        else:
            return self._get(node.right, key)

if __name__ == '__main__':
    bst = BinarySearchTree()

    bst.put("S", 0)
    bst.put("E", 1)
    bst.put("A", 2)
    bst.put("R", 3)
    bst.put("C", 4)
    bst.put("H", 5)
    bst.put("E", 6)
    bst.put("X", 7)
    bst.put("A", 8)
    bst.put("M", 9)
    bst.put("P", 10)
    bst.put("L", 11)
    bst.put("E", 12)

    print("X is at: {}".format(bst.get("X")))
    print("E is at: {}".format(bst.get("E")))

File: datastructures/bst/BinarySearchTree.js
"use strict";

// function Node (key, val, left, right, n=1) {
//     this.key = key;
//     this.val = val;
//     this.left = left;
//     this.right = right;
//     this.n = n;
// }
class Node  {
    constructor(key, val, left, right, n=1) {
        this.key = key;
        this.val = val;
        this.left = left;
        this.right = right;
        this.n = n;
    }
}

class BinarySearchTree {
    constuctor() {
        this.root = null;
    }

    size(node) {
        if (!node) return 0;
        return node.n;
    }

    get(key) {
        return this._get(this.root, key);
    }
    _get(node, key) {
        if (node == null) return -1;
        if (node.key == key) {
            return node.val;
        } else if (key < node.key) {
            return this._get(node.left, key);
        } else {
            return this._get(node.right, key);
        }

    }

    put(key, val) {
        this.root = this._put(this.root, key, val);
    }
    _put(node, key, val) {
        if(!node) return new Node(key, val);
        if(key == node.key) {
            node.val = val
        } else if (key < node.key) {
            node.left = this._put(node.left, key, val);
        } else {
            node.right = this._put(node.right, key, val);
        }

        node.n = this.size(node.left) + this.size(node.right) + 1;
        return node;
    }
}

if (require.main === module) {
    let bst = new BinarySearchTree();
    bst.put("S", 0);
    bst.put("E", 1);
    bst.put("A", 2);
    bst.put("R", 3);
    bst.put("C", 4);
    bst.put("H", 5);
    bst.put("E", 6);
    bst.put("X", 7);
    bst.put("A", 8);
    bst.put("M", 9);
    bst.put("P", 10);
    bst.put("L", 11);
    bst.put("E", 12);

    console.log(`X is at: ${bst.get("X")}`);
    console.log(`E is at: ${bst.get("E")}`);
}
